File: packages/cli/src/commands/generate/dictionary-password-warnings.ts
import type { MigrationPlan } from '@chkit/core'

const PASSWORD_LITERAL_RE = /password\s+'(?!\[HIDDEN\])(?:[^'\\]|\\.)*'/i

export function detectDictionaryPasswordWarnings(plan: MigrationPlan): string[] {
  return plan.operations
    .filter((op) => op.type === 'create_dictionary' && PASSWORD_LITERAL_RE.test(op.sql))
    .map(
      (op) =>
        `Dictionary "${op.key.slice('dictionary:'.length)}" has a password in its SOURCE(...) — it will be written in plain text to the generated migration SQL file.`
    )
}

File: chkit_python/src/chkit/cli/commands/dictionary_password_warnings.py
"""Detect plain-text dictionary passwords in a migration plan.

1:1 port of ``packages/cli/src/commands/generate/dictionary-password-warnings.ts``.
"""

from __future__ import annotations

import re
from typing import Final

from chkit.core.model import MigrationPlan

_PASSWORD_LITERAL_RE: Final[re.Pattern[str]] = re.compile(
    r"password\s+'(?!\[HIDDEN\])(?:[^'\\]|\\.)*'", re.IGNORECASE
)

_DICTIONARY_KEY_PREFIX = "dictionary:"


def detect_dictionary_password_warnings(plan: MigrationPlan) -> list[str]:
    return [
        (
            f'Dictionary "{op.key[len(_DICTIONARY_KEY_PREFIX):]}" has a password in '
            f"its SOURCE(...) — it will be written in plain text to the generated "
            f"migration SQL file."
        )
        for op in plan.operations
        if op.type == "create_dictionary" and _PASSWORD_LITERAL_RE.search(op.sql)
    ]
